File: Firebase/Database/Realtime/FConnection.m
/*
 * Copyright 2017 Google
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "FConnection.h"
#import "FConstants.h"
#import <FirebaseCore/FIRLogger.h>

typedef enum {
    REALTIME_STATE_CONNECTING = 0,
    REALTIME_STATE_CONNECTED = 1,
    REALTIME_STATE_DISCONNECTED = 2,
} FConnectionState;

@interface FConnection () {
    FConnectionState state;
}

@property(nonatomic, strong) FWebSocketConnection *conn;
@property(nonatomic, strong) FRepoInfo *repoInfo;

@end

#pragma mark -
#pragma mark FConnection implementation

@implementation FConnection

@synthesize delegate;
@synthesize conn;
@synthesize repoInfo;

#pragma mark -
#pragma mark Initializers

- (id)initWith:(FRepoInfo *)aRepoInfo
    andDispatchQueue:(dispatch_queue_t)queue
       lastSessionID:(NSString *)lastSessionID {
    self = [super init];
    if (self) {
        state = REALTIME_STATE_CONNECTING;
        self.repoInfo = aRepoInfo;
        self.conn = [[FWebSocketConnection alloc] initWith:self.repoInfo
                                                  andQueue:queue
                                             lastSessionID:lastSessionID];
        self.conn.delegate = self;
    }
    return self;
}

#pragma mark -
#pragma mark Public method implementation

- (void)open {
    FFLog(@"I-RDB082001", @"Calling open in FConnection");
    [self.conn open];
}

- (void)closeWithReason:(FDisconnectReason)reason {
    if (state != REALTIME_STATE_DISCONNECTED) {
        FFLog(@"I-RDB082002", @"Closing realtime connection.");
        state = REALTIME_STATE_DISCONNECTED;

        if (self.conn) {
            FFLog(@"I-RDB082003", @"Calling close again.");
            [self.conn close];
            self.conn = nil;
        }

        [self.delegate onDisconnect:self withReason:reason];
    }
}

- (void)close {
    [self closeWithReason:DISCONNECT_REASON_OTHER];
}

- (void)sendRequest:(NSDictionary *)dataMsg sensitive:(BOOL)sensitive {
    // since this came from the persistent connection, wrap it in a data message
    // envelope
    NSDictionary *msg = @{
        kFWPRequestType : kFWPRequestTypeData,
        kFWPRequestDataPayload : dataMsg
    };
    [self sendData:msg sensitive:sensitive];
}

#pragma mark -
#pragma mark Helpers

- (void)sendData:(NSDictionary *)data sensitive:(BOOL)sensitive {
    if (state != REALTIME_STATE_CONNECTED) {
        @throw [[NSException alloc]
            initWithName:@"InvalidConnectionState"
                  reason:@"Tried to send data on an unconnected FConnection"
                userInfo:nil];
    } else {
        if (sensitive) {
            FFLog(@"I-RDB082004", @"Sending data (contents hidden)");
        } else {
            FFLog(@"I-RDB082005", @"Sending: %@", data);
        }
        [self.conn send:data];
    }
}

#pragma mark -
#pragma mark FWebSocketConnectinDelegate implementation

// Corresponds to onConnectionLost in JS
- (void)onDisconnect:(FWebSocketConnection *)fwebSocket
    wasEverConnected:(BOOL)everConnected {

    self.conn = nil;
    if (!everConnected && state == REALTIME_STATE_CONNECTING) {
        FFLog(@"I-RDB082006", @"Realtime connection failed.");

        // Since we failed to connect at all, clear any cached entry for this
        // namespace in case the machine went away
        [self.repoInfo clearInternalHostCache];
    } else if (state == REALTIME_STATE_CONNECTED) {
        FFLog(@"I-RDB082007", @"Realtime connection lost.");
    }

    [self close];
}

// Corresponds to onMessageReceived in JS
- (void)onMessage:(FWebSocketConnection *)fwebSocket
      withMessage:(NSDictionary *)message {
    NSString *rawMessageType =
        [message objectForKey:kFWPAsyncServerEnvelopeType];
    if (rawMessageType != nil) {
        if ([rawMessageType isEqualToString:kFWPAsyncServerDataMessage]) {
            [self onDataMessage:[message
                                    objectForKey:kFWPAsyncServerEnvelopeData]];
        } else if ([rawMessageType
                       isEqualToString:kFWPAsyncServerControlMessage]) {
            [self onControl:[message objectForKey:kFWPAsyncServerEnvelopeData]];
        } else {
            FFLog(@"I-RDB082008", @"Unrecognized server packet type: %@",
                  rawMessageType);
        }
    } else {
        FFLog(@"I-RDB082009", @"Unrecognized raw server packet received: %@",
              message);
    }
}

- (void)onDataMessage:(NSDictionary *)message {
    // we don't do anything with data messages, just kick them up a level
    FFLog(@"I-RDB082010", @"Got data message: %@", message);
    [self.delegate onDataMessage:self withMessage:message];
}

- (void)onControl:(NSDictionary *)message {
    FFLog(@"I-RDB082011", @"Got control message: %@", message);
    NSString *type = [message objectForKey:kFWPAsyncServerControlMessageType];
    if ([type isEqualToString:kFWPAsyncServerControlMessageShutdown]) {
        NSString *reason =
            [message objectForKey:kFWPAsyncServerControlMessageData];
        [self onConnectionShutdownWithReason:reason];
    } else if ([type isEqualToString:kFWPAsyncServerControlMessageReset]) {
        NSString *host =
            [message objectForKey:kFWPAsyncServerControlMessageData];
        [self onReset:host];
    } else if ([type isEqualToString:kFWPAsyncServerHello]) {
        NSDictionary *handshakeData =
            [message objectForKey:kFWPAsyncServerControlMessageData];
        [self onHandshake:handshakeData];
    } else {
        FFLog(@"I-RDB082012",
              @"Unknown control message returned from server: %@", message);
    }
}

- (void)onConnectionShutdownWithReason:(NSString *)reason {
    FFLog(@"I-RDB082013",
          @"Connection shutdown command received. Shutting down...");

    [self.delegate onKill:self withReason:reason];
    [self close];
}

- (void)onHandshake:(NSDictionary *)handshake {
    NSNumber *timestamp =
        [handshake objectForKey:kFWPAsyncServerHelloTimestamp];
    //    NSString* version = [handshake
    //    objectForKey:kFWPAsyncServerHelloVersion];
    NSString *host = [handshake objectForKey:kFWPAsyncServerHelloConnectedHost];
    NSString *sessionID = [handshake objectForKey:kFWPAsyncServerHelloSession];

    self.repoInfo.internalHost = host;

    if (state == REALTIME_STATE_CONNECTING) {
        [self.conn start];
        [self onConnection:self.conn readyAtTime:timestamp sessionID:sessionID];
    }
}

- (void)onConnection:(FWebSocketConnection *)conn
         readyAtTime:(NSNumber *)timestamp
           sessionID:(NSString *)sessionID {
    FFLog(@"I-RDB082014", @"Realtime connection established");
    state = REALTIME_STATE_CONNECTED;

    [self.delegate onReady:self atTime:timestamp sessionID:sessionID];
}

- (void)onReset:(NSString *)host {
    FFLog(
        @"I-RDB082015",
        @"Got a reset; killing connection to: %@; Updating internalHost to: %@",
        repoInfo.internalHost, host);
    self.repoInfo.internalHost = host;

    // Explicitly close the connection with SERVER_RESET so calling code knows
    // to reconnect immediately.
    [self closeWithReason:DISCONNECT_REASON_SERVER_RESET];
}

@end

File: FirebaseDatabaseSwiftCore/Sources/FConnection.swift
//
//  File.swift
//  File
//
//  Created by Morten Bek Ditlevsen on 14/09/2021.
//

import Foundation

@objc enum FConnectionState : Int {
    case connecting = 0
    case connected = 1
    case disconnected = 2
}

@objc public final class FConnection: NSObject, FWebSocketDelegate {
    @objc public weak var delegate: FConnectionDelegate?
    @objc var state: FConnectionState
    var conn: FWebSocketConnection?
    @objc let repoInfo: FRepoInfo

    @objc public init(
        with aRepoInfo: FRepoInfo,
        andDispatchQueue queue: DispatchQueue,
        googleAppID: String,
        lastSessionID: String?,
        appCheckToken: String?,
        userAgent: String
    ) {
        self.state = .connecting
        self.repoInfo = aRepoInfo
        let connectionURL = aRepoInfo.connectionURL(lastSessionID: lastSessionID)

        conn = FWebSocketConnection(with:
            connectionURL,
            andQueue: queue,
            googleAppID: googleAppID,
            appCheckToken: appCheckToken,
            userAgent: userAgent)

        super.init()

        conn?.delegate = self
    }

    // MARK: -
    // MARK: Public method implementation
    @objc public func open() {
        FFLog("I-RDB082001", "Calling open in FConnection")
        conn?.open()
    }

    @objc public func close(with reason: FDisconnectReason) {
        if state != .disconnected {
            FFLog("I-RDB082002", "Closing realtime connection.")
            state = .disconnected

            if let conn = conn {
                FFLog("I-RDB082003", "Calling close again.")
                conn.close()
                self.conn = nil
            }

            delegate?.onDisconnect(self, withReason: reason)
        }
    }

    @objc public func close() {
        close(with: .DISCONNECT_REASON_OTHER)
    }

    // XXX TODO: Verify that this works
    internal func sendRequestSwift(_ dataMsg: [String: Any?], sensitive: Bool) throws {
        // since this came from the persistent connection, wrap it in a data message
        // envelope
        let msg: [String: Any] = [
            kFWPRequestType: kFWPRequestTypeData,
            kFWPRequestDataPayload: dataMsg
        ]

        try sendData(msg, sensitive: sensitive)
    }


    @objc public func sendRequest(_ dataMsg: NSDictionary, sensitive: Bool) throws {
        // since this came from the persistent connection, wrap it in a data message
        // envelope
        let msg: [String: Any] = [
            kFWPRequestType: kFWPRequestTypeData,
            kFWPRequestDataPayload: dataMsg
        ]

        try sendData(msg, sensitive: sensitive)
    }

    func sendData(_ data: [String: Any], sensitive: Bool) throws {
        if state != .connected {
            /// TODO THROW
//            @throw [[NSException alloc]
//                initWithName:@"InvalidConnectionState"
//                      reason:@"Tried to send data on an unconnected FConnection"
//                    userInfo:nil];
        } else {
            if (sensitive) {
                FFLog("I-RDB082004", "Sending data (contents hidden)")
            } else {
                FFLog("I-RDB082005", "Sending: \(data)")
            }
            self.conn?.send(data)
        }

    }

    // MARK: -
    // MARK: Helpers

    // MARK: -
    // MARK: FWebSocketConnectinDelegate implementation

    // Corresponds to onConnectionLost in JS

    public func onDisconnect(
        _ fwebSocket: AnyObject,
        wasEverConnected everConnected: Bool
    ) {

        self.conn = nil;
        if (!everConnected && state == .connecting) {
            FFLog("I-RDB082006", "Realtime connection failed.")

            // Since we failed to connect at all, clear any cached entry for this
            // namespace in case the machine went away
            repoInfo.clearInternalHostCache()
        } else if (state == .connected) {
            FFLog("I-RDB082007", "Realtime connection lost.")
        }
        self.close()
    }

    // Corresponds to onMessageReceived in JS
    public func onMessage(
        _ fwebSocket: AnyObject,
        withMessage message: [String: Any]
    ) {
        if let rawMessageType = message[kFWPAsyncServerEnvelopeType] as? String {
            if rawMessageType == kFWPAsyncServerDataMessage, let data = message[kFWPAsyncServerEnvelopeData] as? NSDictionary {
                onDataMessage(data)
            } else if rawMessageType == kFWPAsyncServerControlMessage, let data = message[kFWPAsyncServerEnvelopeData] as? [String: Any] {
                onControl(data)
            } else {
                FFLog("I-RDB082008", "Unrecognized server packet type: \(rawMessageType)")
            }
        } else {
            FFLog("I-RDB082009", "Unrecognized raw server packet received: \(message)")
        }
    }

    public func onDataMessage(_ message: NSDictionary?) {
        guard let message = message else { return }
        // we don't do anything with data messages, just kick them up a level
        FFLog("I-RDB082010", "Got data message: \(message)")
        self.delegate?.onDataMessage(self, withMessage: message)
    }

    func onControl(_ message: [String: Any]) {
        FFLog("I-RDB082011", "Got control message: \(message)")
        let type = message[kFWPAsyncServerControlMessageType] as? String
        if type == kFWPAsyncServerControlMessageShutdown, let reason =
            message[kFWPAsyncServerControlMessageData] as? String
        {
            self.onConnectionShutdownWithReason(reason: reason)
        } else if type == kFWPAsyncServerControlMessageReset,
            let host =
                    message[kFWPAsyncServerControlMessageData] as? String {
            self.onReset(host: host)
        } else if type == kFWPAsyncServerHello, let handshakeData =
                    message[kFWPAsyncServerControlMessageData] as? [String: Any] {
            self.onHandshake(handshake: handshakeData)
        } else {
            FFLog("I-RDB082012",
                  "Unknown control message returned from server: \(message)")
        }
    }

    func onConnectionShutdownWithReason(reason: String) {
        FFLog("I-RDB082013",
              "Connection shutdown command received. Shutting down...")

        self.delegate?.onKill(self, withReason: reason)
        self.close()
    }

    func onReset(host: String) {
        FFLog(
            "I-RDB082015",
            "Got a reset; killing connection to: \(repoInfo.internalHost); Updating internalHost to: \(host)")
        self.repoInfo.internalHost = host

        // Explicitly close the connection with SERVER_RESET so calling code knows
        // to reconnect immediately.
        self.close(with: .DISCONNECT_REASON_SERVER_RESET)
    }

    func onHandshake(handshake: [String: Any]) {
        guard let conn = self.conn,
              let timestamp = handshake[kFWPAsyncServerHelloTimestamp] as? NSNumber,
              let host = handshake[kFWPAsyncServerHelloConnectedHost] as? String,
              let sessionID = handshake[kFWPAsyncServerHelloSession] as? String else {
                  // XXX TODO, log error?
                  return
              }

        self.repoInfo.internalHost = host

        if state == .connecting {
            self.conn?.start()
            self.onConnection(conn: conn, readyAtTime: timestamp, sessionID: sessionID)
        }
    }

    func onConnection(conn: FWebSocketConnection, readyAtTime time: NSNumber, sessionID: String) {
        FFLog("I-RDB082014", "Realtime connection established")
        state = .connected
        self.delegate?.onReady(conn, atTime: time, sessionID: sessionID)
    }

}
